SampleApplications/SDK/Opc.Ua.Client/Session.cs

```diff
@@ -3740,6 +3741,16 @@ private void OnKeepAlive(object state)
                     }
                 }
 
+                // limit the number of keep alives sent.
+                int subscriptionCount = SubscriptionCount;
+                int outstandingRequestCount = OutstandingRequestCount;
+                if (outstandingRequestCount > subscriptionCount + m_outstandingRequestThreshold)
+                {
+                    if (!OnKeepAliveError(ServiceResult.Create(StatusCodes.BadNoCommunication, string.Format("Outstanding request threshold has been reached: {0} outstanding requests, {1} subscriptions, {2} configured threshold.", outstandingRequestCount, subscriptionCount, m_outstandingRequestThreshold))))
+                    {
+                        return;
+                    }
+                }
```

Comment:
Hi Salih,

In this place it was a similar code for limiting the number of sent keep-alives:
```
	// limit the number of keep alives sent.
	if (OutstandingRequestCount > SubscriptionCount + 10)
	{
		return;
	}
```
	
I had to remove it because it was causing session reconnect issues in the following scenario:
  - the client connected to a server running on a different machine
  - the client has a subscription without monitored items
  - the network is deactivated on the client machine for a time that generates a keep-alive error (e.g. 20-30 sec.). It only replicates if the network is interupted on client machine!
  - after a successful reconnect by reactivating the existing session on server side, the OutstandingRequestCount object contains some pending Publish requests and causes the OnKeepAlive(object state) method to return without sending a keep-alive read
  - without sending keep-alive read requests KeepAliveStopped returns true and will generate a KeepAliveError and another reconnect and call of ActivateSession in a loop
  
In this case the keep-alive mechanism was stopped by outstanding Publish requests.
  
In your code will have the same issue.

Reply:
Hello Alin,

This piece is added exactly to overcome the situations you listed, plus retaining the ability to be able to trigger a reconnect when there's a routing problem en-route, which we have seen with one of our customers. They have a data center in Canada and a production facility in Germany, connected via a vpn tunnel. during peak times, sometimes the hardware in between just reschedule the packets of a long standing connection causing QoS issues. So, I suggest it is left as I have made the change. The ApplicationConfiguration.cs/xsd files also have the corresponding changes to make the threshold user configurable.

Reply:
Hi Salih,

The problem is that the `m_outstandingRequests` list has an issue and in some cases it ends up with some abandoned Publish requests inside that will remain there during the lifetime of the `Session` object.
One is the case I described above.
Your code change will put the client session in a state where it will constantly trigger KeepAlive errors after each keep alive response until it is recreated. It will also log false error messages with ""KEEP ALIVE LATE:..."
It think is better at least to check in that place only the number of pending KeepAlive **Read** requests. e.g. using
```
        /// <summary>
        /// Gets the number of good outstanding keep-alive requests.
        /// </summary>
        public int GoodKeepAliveRequestCount
        {
            get
            {
                lock (m_outstandingRequests)
                {
                    int count = 0;

                    for (LinkedListNode<AsyncRequestState> ii = m_outstandingRequests.First; ii != null; ii = ii.Next)
                    {
                        if (!ii.Value.Defunct && ii.Value.RequestTypeId == DataTypes.ReadRequest)
                        {
                            count++;
                        }
                    }

                    return count;
                }
            }
        }
```